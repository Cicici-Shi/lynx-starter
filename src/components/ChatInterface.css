/* Copyright 2024 The Lynx Authors. All rights reserved.
 * Licensed under the Apache License Version 2.0 that can be found in the
 * LICENSE file in the root directory of this source tree.
 */

/* ===== 聊天界面主容器 ===== */
.chat-interface {
  display: flex;
  flex-direction: column;
  height: 100%;
  background: var(--rp-bg, #f9fafb);
}

/* ===== 消息滚动区域 ===== */
.chat-messages-scroll {
  flex: 1;
  min-height: 300px;
  padding: 16px;
  overflow-y: auto;
}

/* ===== 消息列表 ===== */
.message-list {
  display: flex;
  flex-direction: column;
  gap: 16px;
  min-height: 100%;
}

.message-wrapper {
  width: 100%;
}

/* ===== 用户消息样式 ===== */
.user-message {
  display: flex;
  justify-content: flex-end;
  margin-bottom: 8px;
}

.user-message .message-content-wrapper {
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  max-width: 80%;
}

.user-message-bubble {
  background: #fcfcfc;
  color: #1d1c1b;
  padding: 12px 16px;
  border-radius: 18px 18px 4px 18px;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.1);
  word-break: break-word;
  white-space: pre-wrap;
}

.user-message-content {
  font-size: 14px;
  line-height: 1.4;
  color: #1d1c1b;
}

.user-message-time {
  font-size: 11px;
  color: var(--rp-text-secondary, #64748b);
  margin-top: 4px;
  margin-right: 4px;
}

/* ===== 机器人消息样式 ===== */
.bot-message {
  display: flex;
  align-items: flex-start;
  margin-bottom: 8px;
}

/* bot-avatar styles removed as per requirements */

.bot-message .message-content-wrapper {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  max-width: 90%;
  flex: 1;
}

/* 为 livestream-selection 消息提供更宽的显示空间 */
.bot-message .message-content-wrapper:has(.livestream-selection) {
  max-width: 95%;
}

/* Lynx 不支持 :has() 伪类，使用特定类名替代 */
.bot-message .message-content-wrapper.livestream-selection-container {
  max-width: 98%;
}

.bot-message-bubble {
  background: var(--rp-panel, #fcfcfc);
  color: #0a0a0a;
  padding: 12px 16px;
  border-radius: 18px 18px 18px 4px;
  border: 1px solid var(--rp-border, #e5e7eb);
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.05);
  word-break: break-word;
  white-space: pre-wrap;
  width: 100%;
}

/* 为包含 livestream-selection 的消息泡泡减少内边距 */
.livestream-selection-container .bot-message-bubble {
  padding: 12px 12px;
}

.bot-message-content {
  font-size: 14px;
  line-height: 1.5;
  color: #0a0a0a;
}

.message-metadata {
  margin-top: 8px;
  padding-top: 8px;
  border-top: 1px solid var(--rp-border, #e5e7eb);
}

.confidence-text {
  font-size: 11px;
  color: var(--rp-text-secondary, #64748b);
}

.bot-message-time {
  font-size: 11px;
  color: var(--rp-text-secondary, #64748b);
  margin-top: 4px;
  margin-left: 4px;
}

/* ===== 思考指示器样式 ===== */
.thinking-indicator-wrapper {
  width: 100%;
}

.thinking-message {
  opacity: 0.8;
}

.thinking-bubble {
  background: var(--rp-panel, #fcfcfc);
  border: 1px solid var(--rp-border, #e5e7eb);
  display: flex;
  align-items: center;
  gap: 12px;
  padding: 16px;
}

.thinking-animation {
  display: flex;
  gap: 4px;
  align-items: center;
}

.thinking-dot {
  width: 6px;
  height: 6px;
  border-radius: 50%;
  background: var(--rp-brand, #10a37f);
  animation: thinking-pulse 1.4s ease-in-out infinite;
}

.thinking-dot.dot-1 {
  animation-delay: 0s;
}

.thinking-dot.dot-2 {
  animation-delay: 0.2s;
}

.thinking-dot.dot-3 {
  animation-delay: 0.4s;
}

.thinking-text {
  font-size: 14px;
  color: var(--rp-text-secondary, #64748b);
  font-style: italic;
}

@keyframes thinking-pulse {
  0%, 60%, 100% {
    transform: scale(1);
    opacity: 0.5;
  }
  30% {
    transform: scale(1.2);
    opacity: 1;
  }
}

/* ===== 消息输入区域 ===== */
.chat-input-container {
  background: var(--rp-panel, #fcfcfc);
  border-top: 1px solid var(--rp-border, #e5e7eb);
  padding: 16px;
  padding-bottom: 30px;
}

.chat-input-wrapper {
  display: flex;
  align-items: flex-end;
  gap: 12px;
  background: var(--rp-bg, #f9fafb);
  border: 1px solid var(--rp-border, #e5e7eb);
  border-radius: 24px;
  padding: 8px 12px;
  transition: border-color 0.2s ease;
}

/* focus-within not supported in Lynx - removed */

.chat-input {
  flex: 1;
  background: transparent;
  border: none;
  outline: none;
  font-size: 14px;
  line-height: 1.4;
  color: var(--rp-text, #0f172a);
  min-height: 20px;
  max-height: 120px;
  padding: 6px 0;
}

.chat-input::placeholder {
  color: var(--rp-text-secondary, #64748b);
}

.chat-input.disabled {
  color: var(--rp-text-secondary, #64748b);
  cursor: not-allowed;
}

.send-button {
  width: 32px;
  height: 32px;
  border-radius: 50%;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  transition: all 0.2s ease;
  flex-shrink: 0;
}

.send-button.active {
  background: var(--rp-brand, #10a37f);
  transform: scale(1);
}

.send-button.active:hover {
  background: var(--rp-brand-hover, #0d8f6b);
  transform: scale(1.05);
}

.send-button.active:active {
  transform: scale(0.95);
}

.send-button.inactive {
  background: var(--rp-border, #e5e7eb);
  cursor: not-allowed;
}

.send-icon.active {
  color: #ffffff;
}

.send-icon.inactive {
  color: var(--rp-text-secondary, #64748b);
}

.input-footer {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-top: 8px;
}

.input-hint {
  text-align: left;
}

.hint-text {
  font-size: 11px;
  color: var(--rp-text-secondary, #64748b);
}

.char-count {
  text-align: right;
}

.char-count-text {
  font-size: 11px;
  color: var(--rp-text-secondary, #64748b);
}

.char-count-text.over-limit {
  color: #ef4444;
  font-weight: 500;
}

.chat-input.over-limit {
  border-color: #ef4444;
  color: #ef4444;
}

/* ===== 响应式设计 ===== */
@media (max-width: 768px) {
  .chat-messages-scroll {
    padding: 12px;
  }
  
  .user-message .message-content-wrapper,
  .bot-message .message-content-wrapper {
    max-width: 90%;
  }
  
  .bot-message .message-content-wrapper.livestream-selection-container {
    max-width: 98%;
  }
  
  .chat-input-container {
    padding: 12px;
  }
  
  .input-hint {
    display: none;
  }
}

/* webkit-scrollbar not supported in Lynx - removed */

/* ===== 动画效果 ===== */
.message-wrapper {
  animation: message-appear 0.3s ease-out;
}

@keyframes message-appear {
  from {
    opacity: 0;
    transform: translateY(10px);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
}

.thinking-indicator-wrapper {
  animation: thinking-appear 0.2s ease-out;
}

@keyframes thinking-appear {
  from {
    opacity: 0;
    transform: translateY(5px);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
}