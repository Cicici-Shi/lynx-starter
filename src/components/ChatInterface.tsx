// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

import { useEffect, useRef } from "@lynx-js/react";

// 声明全局 lynx 对象类型
interface LynxSelectorQuery {
  select(selector: string): {
    invoke(options: {
      method: string;
      params: any;
      success?: (res: any) => void;
      fail?: (res: any) => void;
    }): {
      exec(): void;
    };
  };
}

interface LynxGlobal {
  createSelectorQuery(): LynxSelectorQuery;
}

declare const lynx: LynxGlobal;
import type { ChatInterfaceProps } from "../types/components.ts";
import { MessageList } from "./MessageList.tsx";
import { ChatInput } from "./ChatInput.tsx";

/**
 * 聊天交互组件
 *
 * 功能特性:
 * - 管理聊天消息的展示和输入
 * - 自动滚动到最新消息
 * - 处理用户输入和消息发送
 * - 支持直播记录选择
 * - 显示AI思考状态
 */
export function ChatInterface({
  messages,
  onSendMessage,
  isThinking,
  onRecordSelect,
  onContentUpdate,
}: ChatInterfaceProps) {
  const scrollViewRef = useRef<any>(null);

  // 滚动到底部的函数
  const scrollToBottom = () => {
    try {
      // 使用 Lynx 官方的 SelectorQuery API
      if (typeof lynx !== "undefined" && lynx.createSelectorQuery) {
        lynx
          .createSelectorQuery()
          .select("#chat-messages-scroll")
          .invoke({
            method: "scrollTo",
            params: {
              offset: 999999,
              smooth: true,
            },
          })
          .exec();
      } else {
        // 备用方案：使用 ref 方法
        if (scrollViewRef.current?.scrollToBottom) {
          scrollViewRef.current.scrollToBottom();
        }
      }
    } catch (error) {
      console.warn("自动滚动失败:", error);
    }
  };

  // 监听内容更新，立即滚动
  useEffect(() => {
    if (onContentUpdate) {
      const originalUpdate = onContentUpdate;
      // 如果有内容更新回调，包装它以包含滚动逻辑
    }
  }, [onContentUpdate]);

  // 自动滚动到最新消息
  useEffect(() => {
    // 使用延迟确保DOM更新完成
    const scrollDelay = isThinking ? 150 : 200;

    // 直接使用之前定义的scrollToBottom函数
    const performScroll = scrollToBottom;

    const timeoutId = setTimeout(scrollToBottom, scrollDelay);

    // 清理函数
    return () => {
      clearTimeout(timeoutId);
    };
  }, [messages, isThinking]);

  /**
   * 处理消息发送
   */
  const handleSendMessage = (content: string) => {
    if (!content.trim() || isThinking) {
      return;
    }
    onSendMessage(content.trim());
  };

  /**
   * 处理输入框获得焦点时的滚动
   */
  const handleInputFocus = () => {
    // 延迟滚动，等待软键盘弹起
    setTimeout(() => {
      scrollToBottom();
    }, 350);
  };

  return (
    <view className="chat-interface">
      {/* 消息列表区域 */}
      <scroll-view
        id="chat-messages-scroll"
        ref={scrollViewRef}
        className="chat-messages-scroll"
        scroll-orientation="vertical"
        scroll-bar-enable={true}
        scroll-to-bottom={true}
      >
        <MessageList
          messages={messages}
          isThinking={isThinking}
          onRecordSelect={onRecordSelect}
          onContentUpdate={scrollToBottom}
        />
      </scroll-view>

      {/* 消息输入区域 */}
      <ChatInput
        onSend={handleSendMessage}
        disabled={isThinking}
        placeholder={isThinking ? "Thinking..." : "Check revenue distribution…"}
        onFocus={handleInputFocus}
      />
    </view>
  );
}
