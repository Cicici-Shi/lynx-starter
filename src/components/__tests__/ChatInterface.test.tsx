// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

import "@testing-library/jest-dom";
import { describe, it, expect, vi, beforeEach } from "vitest";
import { render } from "@lynx-js/react/testing-library";
import { ChatInterface } from "../ChatInterface.tsx";
import type { ChatMessage } from "../../types/chatbot.ts";

// Mock the child components
vi.mock("../MessageList.tsx", () => ({
  MessageList: () => <view className="message-list">MessageList</view>,
}));

vi.mock("../ChatInput.tsx", () => ({
  ChatInput: () => <view className="chat-input">ChatInput</view>,
}));

describe("ChatInterface", () => {
  const mockMessages: ChatMessage[] = [
    {
      id: "1",
      type: "user",
      content: "Hello",
      timestamp: new Date("2024-01-01T10:00:00Z"),
    },
    {
      id: "2",
      type: "bot",
      content: "Hi there!",
      timestamp: new Date("2024-01-01T10:01:00Z"),
    },
  ];

  const defaultProps = {
    messages: mockMessages,
    onSendMessage: vi.fn(),
    isThinking: false,
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders chat interface with correct structure", () => {
    render(<ChatInterface {...defaultProps} />);

    // Check if main container exists
    const chatInterface = elementTree.root?.querySelector(".chat-interface");
    expect(chatInterface).toBeTruthy();
  });

  it("renders MessageList component", () => {
    render(<ChatInterface {...defaultProps} />);

    const messageList = elementTree.root?.querySelector(".message-list");
    expect(messageList).toBeTruthy();
  });

  it("renders ChatInput component", () => {
    render(<ChatInterface {...defaultProps} />);

    const chatInput = elementTree.root?.querySelector(".chat-input");
    expect(chatInput).toBeTruthy();
  });

  it("renders scroll view for messages", () => {
    render(<ChatInterface {...defaultProps} />);

    const scrollView = elementTree.root?.querySelector("scroll-view");
    expect(scrollView).toBeTruthy();
  });

  it("applies correct CSS classes", () => {
    render(<ChatInterface {...defaultProps} />);

    expect(elementTree.root?.querySelector(".chat-interface")).toBeTruthy();
    expect(elementTree.root?.querySelector("scroll-view")).toBeTruthy();
  });

  it("should match snapshot", () => {
    render(<ChatInterface {...defaultProps} />);
    expect(elementTree.root).toMatchSnapshot();
  });
});
