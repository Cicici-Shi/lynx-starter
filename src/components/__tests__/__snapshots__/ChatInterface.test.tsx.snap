// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChatInterface > should match snapshot 1`] = `
<page>
  <view
    class="chat-interface"
  >
    <scroll-view
      class="chat-messages-scroll"
      react-ref-3-0="1"
      scroll-bar-enable="true"
      scroll-orientation="vertical"
      scroll-to-bottom="true"
    >
      <view
        class="message-list"
      >
        MessageList
      </view>
    </scroll-view>
    <wrapper>
      <view
        class="chat-input"
      >
        ChatInput
      </view>
    </wrapper>
  </view>
</page>
`;
